---
layout: page
title: "Session 4: Mutation calling and analysis"
parent: Sessions
nav_order: 4
permalink: sessions/session_4/lecture
menubar_toc: true
tabs: session4_tabs
---

<!--script src="{{ site.baseurl }}/assets/js/vanilla-back-to-top.min.js"></script> <script>addBackToTop()</script-->

## Types of somatic genetic variation

## Framework for short variant discovery

The GATK (Genomic Analysis Tool Kit) developed by the Broad Institute is one of the most popular framework/pipelines for processing from raw sequencing reads to identifying short somatic and germline short variants.

{% include image-modal.html link="lecture_assets/1-GATK_logo.png" %}

The workflow is also expanding to include SCNAs and SVs too. In addition, it also includes various tools for processing and performing QC, designed mostly for WES and WGS Illumina data but could be adapted to other technologies.

Their recommended pipeline is called the GATK Best Practices, which provides step-by-step recommendations for processing and analysis from raw reads to variant discovery using high-throughput sequencing. One of the great things about GATK is that it is still being actively developed and supported and is also free to use. There are active forums where users can ask for help or request features.

### The GATK workflow

Here is an overview of the GATK pipeline, consisting of the data pre-processing and the variant discovery steps.

{% include image-modal.html link="lecture_assets/2-GATK_workflow.png" %}

Starting from unmapped reads, it is first necessary to align your reads to the genome and mark duplicate sequences (which we discussed in lecture 3). Once you have aligned and duplicate-marked BAM files you are ready to begin the GATK short somatic variant discovery pipeline depicted here.

GATK has its own somatic variant algorithms called Mutect 2, which can call SNVs and indels. This step usually requires paired tumor/normal samples, and additional information to enable more accurate variant calling such as a Panel of Normals and germline SNP frequencies (more information on these to come).

Mutect2 outputs raw variants that require additional filtering to remove artifacts, germline variants, SNVs from cross-sample contamination, etc. Lastly GATK provides a tool called Funcotator to functionally annotate variants e.g. affected gene, predicted functional consequence, information from databases (e.g. GENCODE, dbSNP, gnomAD, COSMIC), etc. Following this pipeline will produce a set of analysis-ready SNPs which can be further filtered for your specific needs or analyzed directly.

### Pre-processing specifics

In the data pre-processing steps, the raw unmapped reads (in FASTQ or uBAM format) are mapped to a reference genome to generate BAM files. In the GATK workflow these BAM files then undergo additional processing steps called *MarkDuplicatesSpark* and *RecalibrateBaseQualityScores*.

{% include image-modal.html link="lecture_assets/3-GATK_preprocess.png" %}

#### MarkDuplicatesSpark

If you recall from lecture 3, PCR amplification before sequencing can lead to duplicate copies of the same sequencing read, defined as a read that starts and ends at the exact same genomic location. These reads, if not processed, can lead to bias in downstream analyses. For this reason every variant calling workflow should include a step to mark or entirely remove duplicate sequences. Marked duplicate reads are ignored during the variant calling process.

#### RecalibrateBaseQualityScores

Base quality scores are confidence scores generated by the sequencer for each base (for a refresher, review lecture 3). Systematic bias for these scores can originate from library preparation, sequencing, manufacturing defects in flowcell chips, sequencer variation, and sequencing chemistry, resulting in over- or under-estimation of quality scores.

{% include image-modal.html link="lecture_assets/4-RBQS-GATK.png" %}

These scores are important for the variant discovery process as they are used as evidence for or against calling variants. Hence, GATK includes the 'RecalibrateBaseQualityScore' step to correct systematic bias in quality scores. This is accomplished with machine learning to detect and correct for patterns of systematic errors in the base quality scores.

Note that this step, unlike marking duplicates, is optional. In a germline context, recalibrating scores tends to decrease sensitivity in exchange for marginally higher specificity.

### Panel of Normals

{% include image-modal.html link="lecture_assets/5-GATK_PoN.png" %}

The panel of normals refers to a collection of normal samples (ideally >40), which are used to identify technical artifacts from sequencing. For this reason, the best panel of normals is as technically similar to the tumor sequencing samples as possible (i.e. same sample preparation methods, sequencing technology, etc.). To reduce the chances of the panel containing somatic mutations, it's advised to use only blood samples from young individuals with a low chance of undiagnosed cancer.

Some variant callers, including GATK Mutect 2, are able to use the panel of normals in place of a paired normal sample for somatic calling. A paired normal is always better, but this is a good option when you have a tumor with no paired normal. **If you choose to use this option, GATK recommends filtering your somatic variants to only those with a predicted functional consequence** to avoid accidentally calling germline variants as somatic.

### Learn Orientation Bias Artifacts

LearnReadOrientation is a particularly important step and is especially important for FFPE and Illumina Novaseq samples. It helps filter out oxidation and deamination artifacts with higher frequency in one read pair orientation. For instance, if G > T errors always presented in the first Illumina read whereas C > A errors were always found in the second read, this variant would be filtered as an artifact. Read <a href="https://academic.oup.com/nar/article/41/6/e67/2902364" target="_blank">here</a> for more information on read orientation artifacts.

Calculate contamination calculates fraction of reads coming from cross-sample contamination, and this information can be used to filter out additional artifacts.

In this step, F1R2 counts are optional output from Mutect2, which contains the Count of read pairs in the F1R2 and F2R1 configurations supporting the reference and alternate alleles.
F1R2 means first read forward and second read reverse strand. F2R1 means first read (left most) reverse and second read forward  
Conversion of Guanine to 8-oxoG which can pair with C and A
Specific strand orientation in that G > T errors always presented in the first Illumina read, whereas the C > A errors were always found in the second read.


### Cross-sample Contamination



## Variant calling methods and variant callers




## Variant Filtering

## Variant Annotation

## IGV for validation

## Common somatic mutational analyses
