---
layout: page
title: "Session 4: Mutation calling and analysis"
parent: Sessions
nav_order: 4
permalink: sessions/session_4/lecture
menubar_toc: true
tabs: session4_tabs
---

<!--script src="{{ site.baseurl }}/assets/js/vanilla-back-to-top.min.js"></script> <script>addBackToTop()</script-->

---

## Types of somatic genetic variation

---

## Framework for short variant discovery: The GATK workflow

The GATK (Genomic Analysis Tool Kit) developed by the Broad Institute is one of the most popular framework/pipelines for processing from raw sequencing reads to identifying short somatic and germline short variants.

{% include image-modal.html link="lecture_assets/1-GATK_logo.png" %}

The workflow is also expanding to include SCNAs and SVs too. In addition, it also includes various tools for processing and performing QC, designed mostly for WES and WGS Illumina data but could be adapted to other technologies.

Their recommended pipeline is called the GATK Best Practices, which provides step-by-step recommendations for processing and analysis from raw reads to variant discovery using high-throughput sequencing. One of the great things about GATK is that it is still being actively developed and supported and is also free to use. There are active forums where users can ask for help or request features.

Here is an overview of the GATK pipeline, consisting of the data pre-processing and the variant discovery steps.

{% include image-modal.html link="lecture_assets/2-GATK_workflow.png" %}

Starting from unmapped reads, it is first necessary to align your reads to the genome and mark duplicate sequences (which we discussed in lecture 3). Once you have aligned and duplicate-marked BAM files you are ready to begin the GATK short somatic variant discovery pipeline depicted here.

GATK has its own somatic variant algorithms called Mutect 2, which can call SNVs and indels. This step usually requires paired tumor/normal samples, and additional information to enable more accurate variant calling such as a Panel of Normals and germline SNP frequencies (more information on these to come).

Mutect2 outputs raw variants that require additional filtering to remove artifacts, germline variants, SNVs from cross-sample contamination, etc. Lastly GATK provides a tool called Funcotator to functionally annotate variants e.g. affected gene, predicted functional consequence, information from databases (e.g. GENCODE, dbSNP, gnomAD, COSMIC), etc. Following this pipeline will produce a set of analysis-ready SNPs which can be further filtered for your specific needs or analyzed directly.

---

### Pre-processing specifics

In the data pre-processing steps, the raw unmapped reads (in FASTQ or uBAM format) are mapped to a reference genome to generate BAM files. In the GATK workflow these BAM files then undergo additional processing steps called *MarkDuplicatesSpark* and *RecalibrateBaseQualityScores*.

{% include image-modal.html link="lecture_assets/3-GATK_preprocess.png" %}

---

#### MarkDuplicatesSpark

If you recall from lecture 3, PCR amplification before sequencing can lead to duplicate copies of the same sequencing read, defined as a read that starts and ends at the exact same genomic location. These reads, if not processed, can lead to bias in downstream analyses. For this reason every variant calling workflow should include a step to mark or entirely remove duplicate sequences. Marked duplicate reads are ignored during the variant calling process.

---

#### RecalibrateBaseQualityScores

Base quality scores are confidence scores generated by the sequencer for each base (for a refresher, review lecture 3). Systematic bias for these scores can originate from library preparation, sequencing, manufacturing defects in flowcell chips, sequencer variation, and sequencing chemistry, resulting in over- or under-estimation of quality scores.

{% include image-modal.html link="lecture_assets/4-RBQS-GATK.png" %}

These scores are important for the variant discovery process as they are used as evidence for or against calling variants. Hence, GATK includes the 'RecalibrateBaseQualityScore' step to correct systematic bias in quality scores. This is accomplished with machine learning to detect and correct for patterns of systematic errors in the base quality scores.

Note that this step, unlike marking duplicates, is optional. In a germline context, recalibrating scores tends to decrease sensitivity in exchange for marginally higher specificity.

---

### Panel of Normals

{% include image-modal.html link="lecture_assets/5-GATK_PoN.png" %}

The panel of normals refers to a collection of normal samples (ideally >40), which are used to identify technical artifacts from sequencing. For this reason, the best panel of normals is as technically similar to the tumor sequencing samples as possible (i.e. same sample preparation methods, sequencing technology, etc.). To reduce the chances of the panel containing somatic mutations, it's advised to use only blood samples from young individuals with a low chance of undiagnosed cancer.

Some variant callers, including GATK Mutect 2, are able to use the panel of normals in place of a paired normal sample for somatic calling. A paired normal is always better, but this is a good option when you have a tumor with no paired normal. **If you choose to use this option, GATK recommends filtering your somatic variants to only those with a predicted functional consequence** to avoid accidentally calling germline variants as somatic.

---

### Learn Orientation Bias Artifacts

{% include image-modal.html link="lecture_assets/6-GATK_readOrient_contam.png" %}

LearnReadOrientation is a particularly important step and is especially important for FFPE and Illumina Novaseq samples. It helps filter out oxidation and deamination artifacts with higher frequency in one read pair orientation. For instance, if G > T errors always presented in the first Illumina read whereas C > A errors were always found in the second read, this variant would be filtered as an artifact. Read <a href="https://academic.oup.com/nar/article/41/6/e67/2902364" target="_blank">here</a> for more information on read orientation artifacts.

<!-- Image dropdown for read bias artifacts -->

---

### Cross-sample Contamination

Calculate contamination estimates the fraction of reads coming from cross-sample contamination, and this information can be used to filter out additional artifacts. With GATK this is done by searching for reads that match the reference genome at SNP locations where a sample is homozygous for the alternate allele. If the sample is homozygous at that SNP, reads supporting a different allele are presumed to originate from another individual.

---

## Variant calling algorithms

---

### GATK Mutect2

{% include image-modal.html link="lecture_assets/8-mutect2.png" %}

The variant caller within the GATK is Mutect2, which is a consistently top-performing variant calling algorithm (more detailed benchmarks later). Mutect2 has three variant calling modes:

- Paired tumor/normal
- tumor-only mode using the panel of normals
- mitochondria mode

It's important to note that calling mutations in the mitochondrial genome has many unique challenges that require special handling. If calling mitochondrial mutations is your aim, see the distinct <a href="https://github.com/broadinstitute/gatk-docs/blob/master/blog-2012-to-2019/2019-03-05-New!_Mitochondrial_Analysis_with_Mutect2.md?id=23598" target="_blank">GATK mitochondrial variant calling pipeline</a>.

---

### Sentieon variant calling algorithms

We talked in great detail about the GATK pipeline, which is free and widely available. Sentieon, by comparison, is an award-winning commercially developed set of genomics tools (<a href="https://www.sentieon.com/" target="_blank">Sentieon website</a>; <a href="https://www.frontiersin.org/articles/10.3389/fgene.2019.00736/full" target="_blank">published benchmarking paper</a>) which we used for our Sherlock-*Lung* study.

Their pipeline incorporates multiple well known algorithms for DNA and RNA-seq analyses such as BWA-mem, GATK, STAR, Picard, etc. along with their own algorithms such as their somatic variant caller TNscope. Sentieon also includes improvements to existing algorithms such as Sentieon-BWA featuring faster speed, and TNseq which improves upon Mutect2 with increased accuracy. With faster turnaround time, Sentieon algorithms can reduce computing costs and improve variant calling results compared to free pipelines, though of course you will need to pay for a Sentieon license.

---

### Popular Variant Callers

Here we highlight the most popular somatic variant calling algorithms:

| Name                                  | Types of detection | Paired Normal Requirement     | Panel of Normals (PoN) |
|---------------------------------------|--------------------|-------------------------------|------------------------|
| MuTect2                            | SNVs, Indels       | <mark>Somatic tumor-only possible using PoN</mark> | Supported              |
| VarScan2                            | SNVs, Indels       | Required for somatic filtering  | No                     |
| Strelka                             | SNVs, Indels       | Required                      | No                     |
| CaVEman                             | SNVs               | Required                      | Supported              |
| LoFreq                              | SNVs, Indels       | Required for somatic filtering  | No                     |
| VarDict                             | SNVs, Indels       | Required for somatic filtering  | No                     |
| MuSE                                 | SNVs               | Required                      | No                     |
| TnScope/TnSeq (Sentieon, commercial) | SNVs, Indels       | <mark>Somatic tumor-only possible using PoN</mark> | Supported              |
| NeuSomatic                           | SNVs, Indels       | Required                      | No                     |

Most importantly, MuTect2 and the Sentieon algorithms are capable of calling somatic variants without a paired normal by using a panel of normals (PoN), albeit with decreased power to differentiate somatic from germline mutations (thus why GATK recommends filtering variants in this mode to only those with predicted functional consequence). CaVEman is capable of using a PoN, but only for the purposes of filtering, not as a replacement for a paired normal.

Variant calling with Ion Torrent has a high false positive rate, so [an ensemble approach considering results from multiple callers is strongly recommended](https://bmcmedgenomics.biomedcentral.com/articles/10.1186/s12920-019-0636-y), particularly with deep sequencing.

---

### Long-read sequencing

We would also like to briefly discuss long-read sequencing (LRS). <!--Previously, LRS required special algorithms for variant calling for several reasons, most importantly decreased base calling accuracy compared to Illumina. PacBio LRS has since dramatically improved in accuracy to the point that traditional variant callers are theoretically compatible, though benchmarking of their performance on LRS is lacking. For now we suggest using the recommendations from the LRS companies:

<a href="https://www.pacb.com/products-and-services/analytical-software/whole-genome-sequencing/" target="_blank">PacBio's analysis recommendations</a>

<a href="https://nanoporetech.com/sites/default/files/s3/literature/SNVs-and-phasing-workflow.pdf" target="_blank">NanoPore's analysis recommendations</a>-->

---

### Algorithm Benchmarking

<div style="display: flex;
  align-items: center;
  flex-direction: row;">
  <figure>
    {% include image-modal.html max-width="95%" link="lecture_assets/9-variantCaller-PrecisionRecall.png" %}
    <figurecaption class="is-italic">Precision-Recall Curve, stratified by tumor purity (MIX010=10% purity, etc.)</figurecaption>
  </figure>
  <figure>
    {% include image-modal.html max-width="95%" link="lecture_assets/10-variantCaller-memCPUS.png" %}
    <figurecaption class="is-italic">Memory and CPU-hours cost per algorithm</figurecaption>
  </figure>
</div>

Here we include results from one of the most recent benchmarking papers comparing various tools - Mutect2, Strelka2, VarScan2, Neusomatic, and TnScope and TnSeq from Sentieon. On the left hand side is a graph of precision and recall/sensitivity of different callers at various tumor purities/mix ratios (e.g. Higher mixed ratio here means higher tumor purity). As you can see, sample purity substantially affects the precision and recall of all variant callers. Overall, MuTect2 and TnScope consistently have higher recall/sensitivity, particularly at low tumor purity.

In terms of computational cost, Sentieon TnScope and Strelka2 have the lowest resource consumption, which makes Sentieon both highly accurate and resource efficient. So even though Sentieon is not a free software, the cost license could potentially be offset by the lower computational costs.

---

### Ensemble Approaches

Very often, substantial discrepancies exist among the results from different callers. Therefore, most of the time it is recommended to use multiple variant callers in your analysis pipeline. As an example, to this end our very own CGR developed a somatic variant calling pipeline consisting of two parts:

- [Somaticseq2](https://github.com/NCI-CGR/somatic-conda): call somatic variants using different somatic variant callers.
- [Somatic-combiner](https://github.com/mingyi-wang/somatic-combiner): A program to make ensemble calls based upon the output from somaticseq2.

The work has been published in Scientific reports in 2020. The benchmark (using the DREAM dataset) demonstrated that the ensemble calls from somatic-combiner outperform individual callers in general.

{% include image-modal.html link="lecture_assets/11-somaticCombiner.png" %}
*F1 score is the weighted mean of precision and recall.*

---

### NCI Cloud resources for somatic variant calling

Instead of running your somatic variant pipeline on the HPC, there is an emerging interest to run these analyses on the Cloud to eliminate the need to download and store extremely large datasets. Here are a few NCI Cloud resources available for somatic variant calling, as part of the NCI Cancer Research Data Commons that was mentioned in lecture 2. They are based on different cloud platforms so you can choose one that suits your preferences.

- **FireCloud** is a Broad Institute project powered by Terra (which is a cloud platform developed by the Broad Institute, and Microsoft).
- **ISB-CGC** uses Google Cloud Platform so it includes many Google native tools and technologies (including Google BigQuery for big data analytics and Google Compute Engine for complex workflow execution).
- **Cancer Genomics Cloud** is powered by the Seven Bridges platform.

These platforms host large datasets such as TCGA, TARGET, Human Cell Atlas, and have installed several somatic variant calling algorithms/pipelines such as Strelka, VarDict, VarScan, GATK, and possibly others. These resources also host other variant calling algorithms (e.g. germline, SVs, SCNAs) and multi-omics and imaging workflows.

---

### Mutation Calling with Unique Molecular Identifiers (UMIs)



---

## Variant Filtering



---

## Variant Annotation

---

## IGV for validation

---

## Common somatic mutational analyses
