---
layout: page
permalink: sessions/session_2/lecture
menubar_toc: true
---

<script src="{{ site.baseurl }}/assets/js/vanilla-back-to-top.min.js"></script>
<script>addBackToTop()</script>

Download the original slides for this lecture:
- <a target="_blank" href="https://github.com/NCI-ITEB/tumor_epidemiology_approaches_materials/raw/main/lecture_materials/lecture_2/Session%202_%20Public%20Databases%20(presentation).pdf">Lecture slides</a>
- <a target="_blank" href="https://github.com/NCI-ITEB/tumor_epidemiology_approaches_materials/raw/main/lecture_materials/lecture_2/Session%202_%20Public%20Databases%20(Extra%20slides).pdf">Lecture slides+extra materials</a>

## Major Cancer Genomic Studies and Data Portals

- <a target="_blank" href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">TCGA</a>: Hallmark cancer genomics program, containing 33 different tumor types from more than 11,000 patients; analytical results can be accessible from many data portals.
- <a target="_blank" href="https://dcc.icgc.org/">ICGC Data Portal</a>: Cancer genomics data and analysis portal for International Cancer Genome Consortium (ICGC) including over 80 cancer projects, totaling 22 cancer primary sites from over 24,000 donors.
- <a target="_blank" href="https://dcc.icgc.org/pcawg">Pan-Cancer Analysis of Whole Genomes (PCAWG)</a>: Subset of ICGC project focused on whole genome sequencing data and analyses.
- <a target="_blank" href="https://www.genomicsengland.co.uk/initiatives/100000-genomes-project">100,000 Genomes Project \| Genomics England</a>: Sequencing from 100,000 genomes of NHS patients affected by rare diseases and cancer.
- <a target="_blank" href="https://permalinks.stjude.cloud/permalinks/pcgp">Pediatric Cancer Genome Project (PCGP)</a> and <a target="_blank" href="https://pecan.stjude.cloud/">Pediatric Cancer Databases (PeCan)</a>: Databases focused on pediatric cancers, including whole genome, whole exome, and RNA sequencing data.
- <a target="_blank" href="https://platform.stjude.cloud/">St. Jude Cloud Genomics Platform</a>: Cloud genomics platform designed to facilitate NGS analysis.
- <a target="_blank" href="https://proteinpaint.stjude.org/">ProteinPaint</a>: Tool for visualization and analysis of NGS data developed by St. Jude cloud visualization community.
- <a target="_blank" href="https://www.aacr.org/professionals/research/aacr-project-genie/aacr-project-genie-data/">AACR Project: Genomics Evidence Neoplasia Information Exchange (GENIE)</a>: International, open-source, pan-cancer registry of real-world clinical and genomic oncology data; over 136,000 sequencing samples (targeted sequencing).
- <a target="_blank" href="https://www.mskcc.org/msk-impact">MSK-IMPACT</a>: Clinical targeted sequencing cohort using a hybridization capture-based NGS panel at Memorial Sloan Kettering Cancer Center.
- <a target="_blank" href="https://datacommons.cancer.gov/">Cancer Research Data Commons (CRDC) at NCI</a>: NCI cloud-based data science infrastructure that connects data sets with analytics tools to allow users to share, integrate, analyze and visualize cancer research data to drive scientific discovery.
- <a target="_blank" href="https://portal.gdc.cancer.gov/">GDC Data Portal</a>: NCI cancer knowledge network that supports hosting, standardization, and analysis of genomic, clinical, and biospecimen data from cancer research programs.

### Cell Line-Specific Resources

- <a target="_blank" href="https://cellmodelpassports.sanger.ac.uk/">Cell Model Passports</a>: A Hub for Preclinical Cancer Models - Annotation, Genomics & Functional Datasets.
- <a target="_blank" href="https://depmap.org/portal/">Depmap portal</a>: A data portal to empower the research community to make discoveries related to cancer vulnerabilities by providing open access to key cancer dependencies analytical and visualization tools.
- <a target="_blank" href="https://www.cancerrxgene.org/">Genomics of Drug Sensitivity in Cancer</a>: Association between drug response data and genomic markers of sensitivity.

### Advanced Data Portals

These platforms have more advanced data analysis capabilities relative to the portals highlighted previously.

- <a target="_blank" href="https://www.cbioportal.org/datasets">cBioPortal</a>: Most popular cancer genomic data portal. Hosts multi-omics and clinical data from over 300 cancer genomics studies; supports advanced data analysis and programmatic access.
- <a target="_blank" href="https://gdac.broadinstitute.org/">FIREBROWSE</a>: Very useful platform with easy-to-use and advanced genomic analysis methods for TCGA.
- <a target="_blank" href="http://gepia.cancer-pku.cn/index.html">Gene Expression Profiling Interactive Analysis (GEPIA)</a>: Gene expression database and portal for advanced analysis of transcriptomics data including both TCGA and GTEx studies.

## Other Genetics Data Resources

- <a target="_blank" href="https://gtexportal.org/home/">Genotype-Tissue Expression (GTEx)</a>: Public resource to study normal tissue-specific gene expression and regulation. Data type including WGS, WES, bulk and single-nuclear RNAseq.
- <a target="_blank" href="https://gnomad.broadinstitute.org/">gnomAD</a>: Genome aggregation database of with the goal of aggregating and harmonizing both exome and genome sequencing data from a wide variety of large-scale sequencing projects.
- <a target="_blank" href="https://www.ukbiobank.ac.uk">UK Biobank</a>: A large long-term biobank study in the United Kingdom to investigate the respective contributions of genetic predisposition and environmental exposure to the development of disease.
- <a target="_blank" href="https://topmed.nhlbi.nih.gov/">TopMed</a>: Phenotype focused genome sequencing studies, generally focused on blood, heart, and lung genomics and phenotypes.
- <a target="_blank" href="https://www.internationalgenome.org/home">1000 Genomes</a>: Whole-genome sequencing study from healthy populations.
- <a target="_blank" href="https://www.encodeproject.org/">EnCODE and Portal</a>: A public research project with aims to identify functional elements in the human genome.
- <a target="_blank" href="http://www.roadmapepigenomics.org">RoadMap</a>: A public resource of epigenomic maps for stem cells and primary ex vivo tissues selected to represent the normal counterparts of tissues and organ systems frequently involved in human disease.
- <a target="_blank" href="https://epigenomegateway.wustl.edu/browser/">WashU Epigenome Browser</a>: Database allowing for high-level epigenomic topography analysis across regions of the genome.
- <a target="_blank" href="http://3dgenome.fsm.northwestern.edu/index.html">3D Genome Browser</a>: Epigenomics resource particularly useful for topography analysis, similar to WashU Epigenome Browser.
- <a target="_blank" href="https://genome.ucsc.edu/">UCSC Genome Browser</a>: Common resource for genome browsing including many organisms and genomes, and many custom tracks containing specialized annotations.
- <a target="_blank" href="https://useast.ensembl.org/index.html">Ensembl</a>: Genome browser with some helpful tools such as Ensembl Variant Effect Predictor, BLAST, BLAT, BioMart, and more.

## Specialized Databases for Genomic Analyses

- <a target="_blank" href="https://gdac.broadinstitute.org/">FIREBROWSE (Broad GDAC) for Somatic Analysis</a>: Previously mentioned, allows for many very specific cancer genomics analyses for TCGA study.
- <a target="_blank" href="https://dcc.icgc.org/releases/PCAWG">PCAWG Data Portal for WGS-Based Somatic Analysis</a>: Data portal including the major analytical results for PCAWG study.
- <a target="_blank" href="http://compbio.med.harvard.edu/chromothripsis/">Chromothripsis Explorer</a>: Data portal to explore chromothripsis events for PCAWG study.
- <a target="_blank" href="http://dgv.tcag.ca/dgv/app/home?ref=">Database of Genomic Variants (DGV)</a>: Database of germline structural and copy number variation in the human genome.
- <a target="_blank" href="https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/">Ingenuity Pathway Analysis (IPA)</a>: Very useful pathway analysis tool for transcriptomics, proteomics, metabolomics, and more.

### Driver Gene Analysis

- <a target="_blank" href="https://www.intogen.org/search">Integrative Onco Genomics (IntOGen)</a>
- <a target="_blank" href="https://www.oncokb.org/">OncoKB</a>
- <a target="_blank" href="https://cancer.sanger.ac.uk/census">Cosmic Cancer Gene Census</a>
- <a target="_blank" href="https://www.cancerhotspots.org/">Cancer Hotspots</a>

### Alternative Splicing Databases

- <a target="_blank" href="http://www.oncosplicing.com">OncoSplicing</a>
- <a target="_blank" href="http://projects.insilico.us.com/TCGASpliceSeq/">TCGA SpliceSeq</a>
- <a target="_blank" href="http://www.tsvdb.com/index.html">TSVdb</a>

### Mutational Signature Analysis

- <a target="_blank" href="https://cancer.sanger.ac.uk/signatures/">COSMIC Mutational Signatures</a>:  A catalogue of curated reference mutational signatures.
- <a target="_blank" href="https://signal.mutationalsignatures.com">Signal</a>: A web-based tool for exploring cancer and experimentally-generated mutational signatures and performing mutational signature analysis.
- <a target="_blank" href="https://www.ncbi.nlm.nih.gov/research/mutagene/">MUTAGENE</a>: Explore and analyze context-dependent mutational patterns in cancer samples.
- <a target="_blank" href="https://analysistools-dev.cancer.gov/mutational-signatures/#/">mSigPortal</a>: (currently in beta): Integrative mutational signature portal for cancer genomics studies.

## Analytical Programming Packages for Cancer Genomic Datasets

For large scale analysis of many genes at once, programmatic analysis can be extremely useful. Here we highlight some packages for analysis.

- <a target="_blank" href="https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html">TCGAbiolinks</a>: An R/Bioconductor package for integrative analysis with GDC data.
- <a target="_blank" href="https://www.bioconductor.org/packages/release/workflows/html/TCGAWorkflow.html">TCGA Workflow</a>: An integrative package to analyze cancer genomics and epigenomics data using Bioconductor packages with data from TCGA (via GDC), ENCODE, Roadmap.
- <a target="_blank" href="https://bioconductor.org/packages/release/bioc/html/cBioPortalData.html">cBioPortalData</a>: Obtain and analyze data from the cBioPortal API using R.
- <a target="_blank" href="https://github.com/PoisonAlien/maftools">MAFtools</a>: Analysis package for variants files integrating many popular tools for variant analysis, annotation, and visualization.

## Cloud Resources with Available Cancer Genomic Datasets

Cloud resources can be particularly useful for users without access to advanced computing resources.

- <a target="_blank" href="https://terra.bio/">Terra</a>
- <a target="_blank" href="https://www.sevenbridges.com/">Seven Bridges Genomics</a>
- <a target="_blank" href="https://www.dnanexus.com/#">DNANexus</a>

## "Awesome Bioinformatics" Resources

'Awesome Bioinformatics' refers to a collection of introductions for various aspects of bioinformatics analysis. If you're looking for more information on any analysis methods we recommend consulting these documents.

- <a target="_blank" href="https://github.com/broadinstitute/AwesomeGenomics">Awesome genomics</a>: Cancer Data Science's go to place for excellent genomics tools and packages.
- <a target="_blank" href="https://github.com/mikelove/awesome-multi-omics">Awesome multi-omics</a>: A community-maintained list of software packages for multi-omics data analysis.
- <a target="_blank" href="https://github.com/seandavi/awesome-cancer-variant-databases">Awesome cancer variant databases</a>: A community-maintained repository of cancer clinical knowledge bases and databases focused on cancer and normal variants.
- <a target="_blank" href="https://github.com/iron-lion/Awesome-CancerEvolution">Awesome cancer evolution</a>: Papers for studying cancer evolution.
- <a target="_blank" href="https://github.com/cmdcolin/awesome-genome-visualization">Awesome genome visualization</a>: A list of interesting genome visualizers, genome browsers, or genome-browser-like implementations. <a target="_blank" href="https://cmdcolin.github.io/awesome-genome-visualization/?latest=true">New website</a>
- <a target="_blank" href="https://github.com/hammerlab/awesome-clonality">Awesome Clonality</a>: A curated list of awesome resources on clonality and tumor heterogeneity.
- <a target="_blank" href="https://github.com/federicomarini/awesome-expression-browser">Awesome expression browser</a>: A curated list of software and resources for exploring and visualizing (browsing) expression data, but not only limited to that.
- <a target="_blank" href="https://github.com/stevetsa/awesome-microbes">Awesome microbes</a>: List of resources, including software packages (and the people developing these methods) for microbiome (16S), metagenomics (WGS, Shot-gun sequencing), and pathogen identification/detection/characterization.
- <a target="_blank" href="https://github.com/j-andrews7/awesome-bioinformatics-benchmarks">Awesome bioinformatics benchmarks</a>: A curated list of bioinformatics benchmarking papers and resources.
- <a target="_blank" href="https://github.com/seandavi/awesome-single-cell">Awesome single cell</a>: List of software packages (and the people developing these methods) for single-cell data analysis, including RNA-seq, ATAC-seq, etc.
- <a target="_blank" href="https://github.com/danielecook/Awesome-Bioinformatics">Awesome bioinformatics</a>: A curated list of awesome Bioinformatics software, resources, and libraries. Mostly command line based, and free or open-source.
- <a target="_blank" href="https://github.com/crazyhottommy/ChIP-seq-analysis">Awesome ChiP-Seq</a>: A curated list of CHiP-Seq analysis.
- <a target="_blank" href="https://analysistools-dev.cancer.gov/mutational-signatures/#/publications">Awesome mutational signature</a>: A collection of important research papers, reviews, and methods for mutational signature analysis included in <a target="_blank" href="https://analysistools-dev.cancer.gov/mutational-signatures/#/">mSigPortal</a> (currently in beta).
