---
layout: page
permalink: sessions/session_2/lecture
menubar_toc: true
---

<script src="{{ site.baseurl }}/assets/js/vanilla-back-to-top.min.js"></script> <script>addBackToTop()</script>

## Useful Databases

## Major Cancer Genomic Studies and Data Portals

- [TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga): Hallmark cancer genomics program, containing 33 different tumor types from more than 11,000 patients; analytical results can be accessible from many data portals
- [ICGC Data Portal](https://dcc.icgc.org/): Cancer genomics data and analysis portal for International Cancer Genome Consortium (ICGC) including over 80 cancer projects, totaling 22 cancer primary sites from over 24,000 donors.
- [Pan-Cancer Analysis of Whole Genomes (PCAWG)](https://dcc.icgc.org/pcawg): Subset of ICGC project focused on whole genome sequencing data and analyses.
- [100,000 Genomes Project \| Genomics England](https://www.genomicsengland.co.uk/initiatives/100000-genomes-project): Sequencing from 100,000 genomes of NHS patients affected by rare diseases and cancer
- [Pediatric Cancer Genome Project (PCGP)](https://permalinks.stjude.cloud/permalinks/pcgp) and [Pediatric Cancer Databases (PeCan)](https://pecan.stjude.cloud/): Projects and databases focused on pediatric cancers, including whole genome, whole exome, and RNA sequencing data
- [St. Jude Cloud Genomics Platform](https://platform.stjude.cloud/): Cloud genomics platform designed to facilitate the NGS analysis
- [St. Jude Cloud Visualization Community](https://proteinpaint.stjude.org/): Visualization tool for NGS data and analysis developed by St. Jude cloud visualization community
- [AACR Project: Genomics Evidence Neoplasia Information Exchange (GENIE)](https://www.aacr.org/professionals/research/aacr-project-genie/aacr-project-genie-data/): International, open-source, pan-cancer registry of real-world clinical and genomic oncology data; over 136,000 sequencing samples (targeted sequencing)
- [MSK-IMPACT](https://www.mskcc.org/msk-impact): Clinical targeted sequencing cohort using a hybridization capture-based NGS panel at Memorial Sloan Kettering Cancer Center
- [Cancer Research Data Commons (CRDC) at NCI](https://datacommons.cancer.gov/): NCI cloud-based data science infrastructure that connects data sets with analytics tools to allow users to share, integrate, analyze and visualize cancer research data to drive scientific discovery
- [GDC Data Portal](https://portal.gdc.cancer.gov/): NCI cancer knowledge network that supports hosting, standardization, and analysis of genomic, clinical, and biospecimen data from cancer research programs.

### Cell Line-Specific Resources

- [Cell Model Passports](https://cellmodelpassports.sanger.ac.uk/): A Hub for Preclinical Cancer Models - Annotation, Genomics & Functional Datasets
- [Depmap portal](https://depmap.org/portal/): A data portal to empower the research community to make discoveries related to cancer vulnerabilities by providing open access to key cancer dependencies analytical and visualization tools
- [Genomics of Drug Sensitivity in Cancer](https://www.cancerrxgene.org/): Association between drug response data and genomic markers of sensitivity

### Advanced Data Portals

These platforms have more advanced data analysis capabilities relative to the portals highlighted previously.

- [cBioPortal](https://www.cbioportal.org/datasets): Most popular cancer genomic data portal. Hosts multi-omics and clinical data from over 300 cancer genomics studies; supports advanced data analysis and programmatic access
- [FIREBROWSE](https://gdac.broadinstitute.org/): Very useful platform with easy-to-use and advanced genomic analysis methods for TCGA study
- [Gene Expression Profiling Interactive Analysis (GEPIA)](http://gepia.cancer-pku.cn/index.html): Gene expression database and portal for advanced analysis of transcriptomics data including both TCGA and GTEx studies

## Common Genetics Data Resources

- [Genotype-Tissue Expression (GTEx)](https://gtexportal.org/home/): Public resource to study normal tissue-specific gene expression and regulation. Data type including WGS, WES, bulk and single-nuclear RNAseq
- [gnomAD](https://gnomad.broadinstitute.org/): Genome aggregation database of with the goal of aggregating and harmonizing both exome and genome sequencing data from a wide variety of large-scale sequencing projects
- [UK Biobank](https://www.ukbiobank.ac.uk): A large long-term biobank study in the United Kingdom to investigate the respective contributions of genetic predisposition and environmental exposure to the development of disease
- [TopMed](https://topmed.nhlbi.nih.gov/): Phenotype focused genome sequencing studies, generally focused on blood, heart, and lung genomics and phenotypes
- [1000 Genomes](https://www.internationalgenome.org/home): Whole-genome sequencing study from healthy populations
- [EnCODE and Portal](https://www.encodeproject.org/): A public research project with aims to identify functional elements in the human genome
- [RoadMap](http://www.roadmapepigenomics.org): A public resource of epigenomic maps for stem cells and primary ex vivo tissues selected to represent the normal counterparts of tissues and organ systems frequently involved in human disease
- [WashU Epigenome Browser](https://epigenomegateway.wustl.edu/browser/): Database allowing for high-level epigenomic topography analysis across regions of the genome
- [3D Genome Browser](http://3dgenome.fsm.northwestern.edu/index.html): Epigenomics resource particularly useful for topography analysis, similar to WashU Epigenome Browser
- [UCSC Genome Browser](https://genome.ucsc.edu/): Common resource for genome browsing including many organisms and genomes, and many custom tracks containing specialized annotations
- [Ensembl](https://useast.ensembl.org/index.html): Genome browser with some helpful tools such as Ensembl Variant Effect Predictor, BLAST, BLAT, BioMart, and more

## Databases for Specific Genomic Analyses

- [FIREBROWSE (Broad GDAC) for Somatic Analysis](https://gdac.broadinstitute.org/): Previously mentioned, allows for many very specific cancer genomics analyses for TCGA study
- [PCAWG Data Portal for WGS-Based Somatic Analysis](https://dcc.icgc.org/releases/PCAWG): Data portal including the major analytical results for PCAWG study
- [Chromothripsis Explorer](http://compbio.med.harvard.edu/chromothripsis/): Data portal to explore the chromothripsis events for PCAWG study
- [Database of Genomic Variants (DGV)](http://dgv.tcag.ca/dgv/app/home?ref=): Database of germline structural and copy number variation in the human genome
- [Ingenuity Pathway Analysis (IPA)](https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/): Very useful pathway analysis tool for transcriptomics, proteomics, metabolomics, and more

### Driver Gene Analysis

- [Integrative Onco Genomics (IntOGen)](https://www.intogen.org/search)
- [OncoKB](https://www.oncokb.org/)
- [Cosmic Cancer Gene Census](https://cancer.sanger.ac.uk/census)
- [Cancer Hotspots](https://www.cancerhotspots.org/)

### Alternative Splicing Databases

- [OncoSplicing](http://www.oncosplicing.com)
- [TCGA SpliceSeq](http://projects.insilico.us.com/TCGASpliceSeq/)
- [TSVdb](http://www.tsvdb.com/index.html)

### Mutational Signature Analysis

- [COSMIC Mutational Signatures](https://cancer.sanger.ac.uk/signatures/):  A catalogue of curated reference mutational signatures
- [Signal](https://signal.mutationalsignatures.com): A web-based tool for exploring cancer and experimentally-generated mutational signatures and performing mutational signature analysis
- [MUTAGENE](https://www.ncbi.nlm.nih.gov/research/mutagene/): Explore and analyze context-dependent mutational patterns in cancer samples
- mSigPortal (coming soon): Integrative mutational signature portal for cancer genomics studies

## Analytical Programming Packages for Cancer Genomic Datasets

For large scale analysis of many genes at once, programmatic analysis can be extremely useful. Here we highlight some packages for analysis.

- [TCGAbiolinks](https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html): R package for
- [TCGA Workflow](https://www.bioconductor.org/packages/release/workflows/html/TCGAWorkflow.html): An integrative package to analyze cancer genomics and epigenomics data using Bioconductor packages with data from TCGA (via GDC), ENCODE, Roadmap
- [cBioPortalData](https://bioconductor.org/packages/release/bioc/html/cBioPortalData.html): Obtain and analyze data from the cBioPortal API using R
- [MAFtools](https://github.com/PoisonAlien/maftools): Analysis package for variants files integrating many popular tools for variant analysis, annotation, and visualization

## Cloud Resources with Available Cancer Genomic Datasets

Cloud resources can be particularly useful for users without access to advanced computing resources.

- [Terra](https://terra.bio/)
- [Seven Bridges Genomics](https://www.sevenbridges.com/)
- [DNANexus](https://www.dnanexus.com/#)

## "Awesome Bioinformatics" Resources

'Awesome Bioinformatics' refers to a collection of introductions for various aspects of bioinformatics analysis. If you're looking for more information on any analysis methods we recommend consulting these documents.

- [Awesome genomics](https://github.com/broadinstitute/AwesomeGenomics): Cancer Data Science's go to place for excellent genomics tools and packages.
- [Awesome multi-omics](https://github.com/mikelove/awesome-multi-omics): A community-maintained list of software packages for multi-omics data analysis.
- [Awesome cancer variant databases](https://github.com/seandavi/awesome-cancer-variant-databases): A community-maintained repository of cancer clinical knowledge bases and databases focused on cancer and normal variants
- [Awesome cancer evolution](https://github.com/iron-lion/Awesome-CancerEvolution): Papers for studying cancer evolution
- [Awesome genome visualization](https://github.com/cmdcolin/awesome-genome-visualization): A list of interesting genome visualizers, genome browsers, or genome-browser-like implementations. [New website.](https://cmdcolin.github.io/awesome-genome-visualization/?latest=true)
- [Awesome Clonality](https://github.com/hammerlab/awesome-clonality): A curated list of awesome resources on clonality and tumor heterogeneity.
- [Awesome expression browser](https://github.com/federicomarini/awesome-expression-browser): A curated list of software and resources for exploring and visualizing (browsing) expression data, but not only limited to that.
- [Awesome microbes](https://github.com/stevetsa/awesome-microbes): List of resources, including software packages (and the people developing these methods) for microbiome (16S), metagenomics (WGS, Shot-gun sequencing), and pathogen identification/detection/characterization.
- [Awesome bioinformatics benchmarks](https://github.com/j-andrews7/awesome-bioinformatics-benchmarks): A curated list of bioinformatics benchmarking papers and resources.
- [Awesome single cell](https://github.com/seandavi/awesome-single-cell): List of software packages (and the people developing these methods) for single-cell data analysis, including RNA-seq, ATAC-seq, etc.
- [Awesome bioinformatics](https://github.com/danielecook/Awesome-Bioinformatics): A curated list of awesome Bioinformatics software, resources, and libraries. Mostly command line based, and free or open-source.
- [Awesome ChiP-Seq](https://github.com/crazyhottommy/ChIP-seq-analysis): A curated list of CHiP-Seq analysis
- Awesome mutational signature (coming soon on mSigPortal): A collection of important research papers, reviews, and methods for mutational signature analysis to be launched on mSigPortal
